arrows/core/metadata_map_io_csv.cxx

```diff
@@ -192,18 +277,68 @@ ::save_( std::ostream& fout,
         auto const type_id = metadata_item.first;
         if( type_id != kv::VITAL_META_VIDEO_URI )
         {
-          metadata_ids.insert( type_id );
+          present_metadata_ids.insert( type_id );
         }
       }
     }
   }
 
+  std::vector< std::string > metadata_names;
+  std::vector< kv::vital_metadata_tag > ordered_metadata_ids;
+
+  for( auto const& name : d_->column_names )
+  {
+    auto trait_id = d_->md_traits.enum_name_to_tag( name );
+    if( trait_id != kv::VITAL_META_UNKNOWN )
+    {
+      // This is a placeholder to keep the two vectors aligned
+      metadata_names.push_back( "" );
+      // Avoid duplicating present columns
+      present_metadata_ids.erase( trait_id );
+    }
+    else if( kv::VITAL_META_UNKNOWN !=
+             ( trait_id = d_->md_traits.name_to_tag( name ) ) )
+    {
+      // This is a placeholder to keep the two vectors aligned
+      metadata_names.push_back( "" );
+      // Avoid duplicating present columns
+      present_metadata_ids.erase( trait_id );
+      LOG_INFO(
+        logger(),
+        "Description "  << name << " matched enum "
+                        << d_->md_traits.tag_to_enum_name( trait_id ) );
+    }
+    else
+    {
+      // TODO Consider whether UNKNOWN is the right tag or if something to show
+      // explicitly that this is not in our set of tags is better
+      metadata_names.push_back( name );
+    }
+    ordered_metadata_ids.push_back( trait_id );
+  }
+
+  // Determine whether to write columns present in the metadata but not
+  // explicitly provided
+  if( d_->write_remaining_columns )
+  {
+    for( auto const& id : present_metadata_ids )
+    {
+      ordered_metadata_ids.push_back( id );
+      // Again, this is just a placeholder to keep vectors aligned
+      metadata_names.push_back( "" );
+    }
+  }
+
   // Write out the csv header
   fout << "\"frame ID\",";
-  for( auto const& metadata_id : metadata_ids )
+  assert( ordered_metadata_ids.size() == metadata_names.size() );
+  for( auto const& i : iota( ordered_metadata_ids.size() ) )
   {
-    d_->write_csv_header( metadata_id, fout );
+    auto const& metadata_id = ordered_metadata_ids.at( i );
```

Comment:
Not a big deal, but you could probably just use `[i]` here instead of `.at( i )`.  I think the only difference is `.at` has bounds checking, which you shouldn't need when iterating through the items like this.